awx/main/tasks/receptor.py

```diff
@@ -266,139 +267,119 @@ def __init__(self, task, runner_params=None):
             self.runner_params['only_transmit_kwargs'] = True
 
     def run(self):
-        # We establish a connection to the Receptor socket
-        receptor_ctl = get_receptor_ctl()
-
         res = None
+        unit_id = None
         try:
-            res = self._run_internal(receptor_ctl)
+            res, unit_id = self._run()
+
+            if res.status == 'error':
+                self.handle_error(unit_id)
+
             return res
         finally:
             # Make sure to always release the work unit if we established it
-            if self.unit_id is not None and settings.RECEPTOR_RELEASE_WORK:
+            if unit_id is not None and settings.RECEPTOR_RELEASE_WORK:
                 try:
-                    receptor_ctl.simple_command(f"work release {self.unit_id}")
+                    self.receptor_ctl.simple_command(f"work release {unit_id}")
                 except Exception:
-                    logger.exception(f"Error releasing work unit {self.unit_id}.")
+                    logger.exception(f"Error releasing work unit {unit_id}.")
+
+    def _run(self):
+        payload_writer, payload_reader = socket.socketpair()
+
+        # This ThreadPoolExecutor runs for the duration of the job.
+        # The cancel_func pattern is intended to guard againsnt any situation where we may
+        # end up stuck while reading or writing from the receptor socket.
+        with concurrent.futures.ThreadPoolExecutor(max_workers=3) as executor:
```

Comment:
I checked out your branch and ran graphs

![Screenshot from 2022-08-16 08-32-44](https://user-images.githubusercontent.com/1385596/184880592-22585fee-df2a-4db6-9d57-12b8c50b66dc.png)

This is for running 5 jobs. The increase in database connections is 30-10 = 20. Then 20 / 5 = 4 threads per job, as expected. This is because `max_workers=3` doesn't include the main thread, so 4 threads in total, and without connection pooling between threads, a new connection is created for each. Currently it's 3 per job, and particularly if we're backporting, increasing this number by 1 will be disastrous for some large (or poorly balanced) deployments.

Really, it should be either 1 or 0. But for now, I'm going to pull out some of my tricks to suggest to mitigate what you're doing here, because watching for cancels over the transmit phase is something I know we were lacking.